utils/errors.go

```diff
@@ -0,0 +1,73 @@
+package utils
+
+import (
+	"fmt"
+	"time"
+
+	awsDynamoTypes "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/rs/zerolog"
+	"github.com/segmentio/kafka-go"
+)
+
+// ProcessingError is an error used for Kafka processing that communicates retry data for
+// failures.
+type ProcessingError struct {
+	OriginalError  error
+	FailureMessage string
+	Temporary      bool
+	Backoff        time.Duration
+	KafkaMessage   kafka.Message
+}
+
+// Error makes ProcessingError an error
+func (e ProcessingError) Error() string {
+	msg := fmt.Sprintf("error: %s", e.FailureMessage)
+	if e.OriginalError != nil {
+		msg = fmt.Sprintf("%s: %s", msg, e.OriginalError)
+	}
+	return msg
+}
+
+// Cause implements Cause for error
+func (e ProcessingError) Cause() error {
+	return e.OriginalError
+}
+
+// ProcessingErrorFromErrorWithMessage converts an error into a ProcessingError
+func ProcessingErrorFromErrorWithMessage(
+	err error,
+	message string,
+	kafkaMessage kafka.Message,
+	logger *zerolog.Logger,
+) *ProcessingError {
+	temporary, backoff := ErrorIsTemporary(err, logger)
+	return &ProcessingError{
+		OriginalError:  err,
+		FailureMessage: message,
+		Temporary:      temporary,
+		Backoff:        backoff,
+		KafkaMessage:   kafkaMessage,
+	}
+}
+
+// ErrorIsTemporary takes an error and determines
+func ErrorIsTemporary(err error, logger *zerolog.Logger) (bool, time.Duration) {
+	var ok bool
+	err, ok = err.(*awsDynamoTypes.ProvisionedThroughputExceededException)
```

Comment:
Should we use `errors.As` in case the error is wrapped by our code instead of straight up type checking?

Reply:
I think I got this sorted.